language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -421,10 +525,155 @@ genModuleRef (pkgRef, modName) = case pkgRef of
+    , "  \"scripts\": {"
+    , "    \"build\": \"tsc --build\","
+    , "    \"lint\": \"eslint --ext .ts src/ --max-warnings 0\""
+    , "  },"
+    , "  \"devDependencies\": {"
+    , "    \"@typescript-eslint/eslint-plugin\": \"^2.11.0\","
+    , "    \"@typescript-eslint/parser\": \"^2.11.0\","
+    , "    \"eslint\": \"^6.7.2\","
+    , "    \"typescript\": \"~3.7.3\""
+    , "  }"
+    , "}"
+    ])
+  where
+    packageName =  packageNameOfPackageDir packageDir
+    dependencies = withCommas [ "    \"" <> pkg <> "\": \"" <> packageVersion <> "\""
+                              | d <- depends
+                              , let pkg = scope ++ "/" ++ undependency d
+                              ]
+
+    withCommas :: [String] -> [String]
+    withCommas [] = []
+    withCommas ms = reverse (head ms' : map (++",") (tail ms')) where ms' = reverse ms
+
+-- This type describes the format of a "top-level" 'package.json'. We
+-- expect such files to have the format
+-- {
+--   "private": true,
+--   "workspaces: [
+--      "path/to/foo",
+--      "path/to/bar",
+--      ...
+--   ]
+-- }
+data PackageJson = PackageJson
+  {  private :: Bool
+  ,  workspaces :: [T.Text]
+  } deriving Show
+-- Explicitly provide instances to avoid relying on restricted
+-- extension 'DeriveGeneric'.
+instance FromJSON PackageJson where
+  parseJSON (Object v) = PackageJson
+      <$> v .: "private"
+      <*> v .: "workspaces"
+  parseJSON _ = mzero
+instance ToJSON PackageJson where
+  toJSON (PackageJson private workspaces) =
+    object [ "private" .= private
+           , "workspaces" .= workspaces
+           ]
+
+-- Read the provided 'package.json'; transform it to include the
+-- provided workspaces; write it back to disk.
+writeTopLevelPackageJson :: FilePath -> [(String, String, [Dependency])] -> FilePath -> IO ()
+writeTopLevelPackageJson optOutputDir dependencies file = do
+  let (g, nodeFromVertex) = graphFromEdges'
+        (map (\(a, b, ds) -> (a, b, map undependency ds)) dependencies)
+      ps = map (fst3 . nodeFromVertex) $ reverse (topSort g)
+        -- Topologically order our packages.
+      ldr = fromJust (stripPrefix "@" (unscope (scopeOfScopeDir optOutputDir)))
+        -- 'ldr' we expect to be something like "daml2ts/".
+  let ourPackages = map (T.pack . ((ldr ++ "/") ++)) ps
+  bytes <- BSL.readFile file
+  maybe
+    (fail $ "Error decoding JSON from '" <> file <> "'")
+    (transformAndWrite ourPackages ldr)
+    (decode bytes :: Maybe PackageJson)
+  where
+    transformAndWrite :: [T.Text] -> String -> PackageJson -> IO ()
+    transformAndWrite ourPackages ldr oldPackageJson = do
+      let damlTypes = [T.pack "daml-types"]
+      --   * Old versions of our packages should be removed;
+          keepPackages =
+            [ T.pack x
+            | x <- [T.unpack y | y <- workspaces oldPackageJson]
+            , isNothing $ stripPrefix ldr x
+            ]
+      --  * Our packages need to come after 'daml-types' if it exists;
```

Comment:
We can remove the special case for `daml-types`, which is actually `@daml/types` once https://github.com/digital-asset/daml/pull/4820 has landed. The only reason why you would need this is when you don't get `@daml/types` from npmjs.com but rather have it somehow local, as we do in our tests. However, we can also use the `resolutions` field in the workspace `package.json` to point to our local version of `@daml/types`, as in https://github.com/digital-asset/daml/pull/4820.

Reply:
In other words, the packages generated by `daml2ts` go first, everything else comes afterwards. This should simplify the code here quite a bit.

Reply:
Yes, got it. Will implement.

Reply:
https://github.com/digital-asset/daml/pull/4859